### ISSUE TITLE
Bootsnap breaks Zeitwerk on Ruby 2.5.x

### ISSUE THREAD

**Issue by casperisfine** [Shopify/bootsnap#244]
There's a bug on Ruby 2.5, when code is loaded a ISeq TracePoint callbacks are not fired:

```ruby
iseq = RubyVM::InstructionSequence.compile(<<-RUBY)
  class A
    class B
    end
  end
RUBY

iseq_bin = iseq.to_binary
TracePoint.new(:end){|tp|
  p tp.lineno
}.enable{
  RubyVM::InstructionSequence.load_from_binary(iseq_bin).eval
}

TracePoint.new(:class){|tp|
  p "class #{tp.lineno}"
}.enable{
  RubyVM::InstructionSequence.load_from_binary(iseq_bin).eval
}
```

This snippet prints nothing on MRI 2.5, but works as expected on 2.6.

The issue with this is that Zeitwerk, that's the default autoloader on Rails 6 rely on TracePoint to work properly. Since Rails also include bootsnap as default, 2.5 users trying to use Rails 6 might encounter very weird load issues.

I'm not quite sure what can be done, but It would probably be best to figure something out before Rails 6 is released.

@burke @fxn @rafaelfranca thoughts on this?

Rails load `bootsnap/setup`: https://github.com/Shopify/bootsnap/blob/master/lib/bootsnap/setup.rb

I suppose we could check the Ruby version and set `compile_cache_iseq: false` for MRI 2.5 ?

The perf hit is quite significant, but if it can save headaches for users, it might be best.

**Comment by rafaelfranca** [comment]
This bug is fixed already in Ruby, it is just not released. They are planning a release next week. Related with https://github.com/rails/rails/issues/35475.

I think we can just wait the new Ruby release. It will happen before Rails 6 final is released.

**Comment by casperisfine** [comment]
Oh great. Well closing this then.

**Comment by fxn** [comment]
Ruby 2.5.5 still has this problem.

**Comment by casperisfine** [comment]
Yeah we discovered this the hard way. The TracePoint patch was supposed to be part of 2.5.5, but the release was borked and that fix wasn't properly included...

**Comment by fxn** [comment]
For the archives: after https://github.com/Shopify/bootsnap/pull/257 the default setup has this cache disabled.

The issue is probably worth being open still, however, because it affects people enabling the cache by hand.

**Comment by casperisfine** [comment]
Seems like the fix was never properly backported to 2.5, however bootsnap no longer enable ISeq caching by default on 2.5.x, and I added a warning if users enable it manually: https://github.com/Shopify/bootsnap/commit/eab8d8df45e049ac7a5e35a8fb7cd403ff07298c

I think we can close this now.
